File: ws-client/src/main/java/net/devopssolutions/demo/ws/client/component/WsConsumer.java
package net.devopssolutions.demo.ws.client.component;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.glassfish.tyrus.client.ClientManager;
import org.glassfish.tyrus.container.jdk.client.JdkClientContainer;
import org.glassfish.tyrus.core.CloseReasons;
import org.glassfish.tyrus.ext.client.java8.SessionBuilder;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import javax.annotation.PreDestroy;
import javax.websocket.*;
import java.io.IOException;
import java.net.URI;
import java.nio.ByteBuffer;
import java.util.concurrent.atomic.AtomicReference;

@Slf4j
@Component
public class WsConsumer {

    private final ClientManager clientManager = ClientManager.createClient(JdkClientContainer.class.getName());
    private final SessionBuilder sessionBuilder = new SessionBuilder(clientManager)
            .uri(new URI("ws://localhost:8080/ws"))
            .onOpen(this::onOpen)
            .messageHandler(PongMessage.class, new MessageHandler.Whole<PongMessage>() {//bug in jdk
                @Override
                public void onMessage(PongMessage pongMessage) {
                    WsConsumer.this.onPong(pongMessage);
                }
            })
            .messageHandler(ByteBuffer.class, this::onMessage)
            .messageHandler(String.class, this::onMessage)
            .onError(this::onError)
            .onClose(this::onClose);

    @Getter
    private final AtomicReference<Session> session = new AtomicReference<>();
    private final ByteBuffer ping = ByteBuffer.allocate(0);

    public WsConsumer() throws Exception {
        log.info("going to open ws connection");
        connect();
    }

    private boolean connect() {
        try {
            this.session.set(sessionBuilder.connect());
            return true;
        } catch (Exception e) {
            this.session.set(null);
            log.warn("couldn't connect", e);
            return false;
        }
    }

    private void onPong(PongMessage pongMessage) {
        log.info("incoming onPong: {}", pongMessage);
    }

    private void onMessage(ByteBuffer message) {
        log.info("new message: {}", message);
    }

    private void onMessage(String message) {
        log.info("new message: {}", message);
    }

    private void onOpen(Session session, EndpointConfig endpointConfig) {
        log.info("opened ws connection");
    }

    private void onClose(Session session, CloseReason closeReason) {
        log.info("closing ws connection");
    }

    private void onError(Session session, Throwable throwable) {
        log.warn("error in ws session: " + session, throwable);
    }

    @Scheduled(fixedRate = 10_000)
    private void sendPing() {
        Session session = this.session.get();
        log.info("sending ping: {}", session != null);
        if (session != null) {
            try {
                session.getBasicRemote().sendPing(ping);
            } catch (Exception e) {
                this.session.set(null);
                log.warn("exception sending ping", e);
            }
        } else {
            connect();
        }
    }

    @PreDestroy
    private void stop() {
        Session session = this.session.get();
        if (session != null && session.isOpen()) {
            try {
                session.close(CloseReasons.GOING_AWAY.getCloseReason());
            } catch (IOException e) {
                log.warn("exception closing session", e);
            }
        }
    }

}

File: ws-client/src/main/java/net/devopssolutions/demo/ws/client/component/WsConsumer.kt
package net.devopssolutions.demo.ws.client.component

import mu.KLogging
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import org.springframework.web.reactive.socket.CloseStatus
import org.springframework.web.reactive.socket.WebSocketMessage
import org.springframework.web.reactive.socket.WebSocketSession
import org.springframework.web.reactive.socket.client.WebSocketClient
import reactor.core.Disposable
import reactor.core.publisher.EmitterProcessor
import reactor.core.publisher.Flux
import reactor.core.publisher.FluxSink
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import java.net.URI
import java.time.Duration
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.atomic.AtomicReference
import javax.annotation.PreDestroy

@Component
open class WsConsumer(
        private val webSocketClient: WebSocketClient,
        private val wsProducer: WsProducer
) {
    companion object : KLogging()

    private val messagesCount = AtomicLong(0)
    private val messagesCountInSecond = AtomicLong(0)
    private val subscriptionReference = AtomicReference<Disposable?>()

    private fun connect() {
        logger.info("connecting ...")
        val subscription = webSocketClient
                .execute(URI("ws://localhost:8080/ws"), this::handle)
                .subscribe()
        if (subscriptionReference.compareAndSet(null, subscription)) {
            messagesCount.set(0)

            logger.info("subscription set")
        } else {
            logger.warn("another session is connected")
            subscription.dispose()
        }
    }

    private fun handle(session: WebSocketSession): Mono<Void> {
        logger.info("opened ws connection session: {}, endpointConfig: {}", session, session.handshakeInfo)

        val emitterProcessor = EmitterProcessor.create<WebSocketMessage>(1000, false)
        val sink = emitterProcessor.sink(FluxSink.OverflowStrategy.BUFFER)
        buildPinger(session).subscribe { sink.next(it) }
        wsProducer.sendHello(session).subscribe { sink.next(it) }
        wsProducer.sendFlood(session).subscribe { sink.next(it) }

        return session.send(emitterProcessor.subscribeOn(Schedulers.newParallel("foo", 4)))
                .mergeWith(createReceiver(session, sink).subscribeOn(Schedulers.newParallel("bar", 4)).then())
                .then()
                .doOnError { exception -> afterConnectionClosed(session, CloseStatus.SERVER_ERROR, exception) }
                .doOnSuccess { afterConnectionClosed(session, CloseStatus.GOING_AWAY) }
                .doOnCancel {
                    logger.warn("will close session")
                    session.close().block()
                }
                .doOnSubscribe { logger.info("starting {}", it) }
    }

    private fun createReceiver(session: WebSocketSession, sink: FluxSink<WebSocketMessage>)
            : Flux<WebSocketMessage> = session.receive().doOnNext { message -> handleMessage(session, sink, message) }

    private fun buildPinger(session: WebSocketSession): Flux<WebSocketMessage> = Flux.interval(Duration.ofSeconds(30))
            .map { session.pingMessage { it.allocateBuffer(0) } }
            .doOnNext { logger.info("sending ping") }

    private fun afterConnectionClosed(session: WebSocketSession?, closeStatus: CloseStatus?, exception: Throwable? = null) {
        logger.info("ws connection closed, session: {}, closeReason: {}", session, closeStatus, exception)
    }

    private fun handleMessage(
            session: WebSocketSession,
            sink: FluxSink<WebSocketMessage>,
            message: WebSocketMessage) = when (message.type) {
        WebSocketMessage.Type.PING -> handlePingMessage(session, sink, message)
        WebSocketMessage.Type.TEXT -> handleTextMessage(session, message)
        WebSocketMessage.Type.BINARY -> handleBinaryMessage(session, message)
        WebSocketMessage.Type.PONG -> handlePongMessage(session, message)
    }

    private fun handlePingMessage(session: WebSocketSession, sink: FluxSink<WebSocketMessage>, message: WebSocketMessage) {
        logger.info("incoming Ping: {} will respond with pong", message)
        sink.next(session.pongMessage { it.allocateBuffer(0) })
    }

    private fun handlePongMessage(session: WebSocketSession, message: WebSocketMessage) {
        logger.info("incoming Pong: {}", message)
    }

    private fun handleTextMessage(session: WebSocketSession, message: WebSocketMessage) {
        logger.info("onTextMessage length: {}, message: {}", message.payload.readableByteCount(), message.payloadAsText)
    }

    private fun handleBinaryMessage(session: WebSocketSession, message: WebSocketMessage) {
        val count = messagesCount.incrementAndGet()
        messagesCountInSecond.incrementAndGet()
        if (count % 100_000 == 1L) {
            logger.info("onBinaryMessage count: {}, message: {}, length: {}", count, message, message.payload.readableByteCount())
        }
    }

    @Scheduled(fixedRate = 1_000)
    private fun countMessages() {
        val count = messagesCountInSecond.getAndSet(0)
        logger.info("messagesCountInSecond: {}", count)
    }

    @Scheduled(fixedRate = 5_000)
    private fun reConnect() {
        val disposable = this.subscriptionReference.updateAndGet {
            if (it?.isDisposed == true) null else it
        }
        if (disposable == null) {
            connect()
        }
    }

    @PreDestroy
    private fun stop() {
        subscriptionReference.getAndSet(null)?.dispose()
    }

}
